src/core_editor/editor.rs

```diff
@@ -125,20 +147,66 @@ impl Editor {
         self.line_buffer.is_empty()
     }
 
-    pub fn undo(&mut self) -> Option<()> {
-        self.line_buffer.undo()
+    pub fn reset_olds(&mut self) {
+        self.edits = vec![LineBuffer::new()];
+        // NOTE: Leaving the old logic to help with the review
+        // self.old_lines = vec![vec![String::new()]];
+        // self.old_insertion_point = vec![InsertionPoint::new()];
+        self.index_undo = 2;
+    }
+
+    fn get_index_undo(&self) -> usize {
+        if let Some(c) = self.edits.len().checked_sub(self.index_undo) {
+            c
+        } else {
+            0
+        }
     }
 
     pub fn redo(&mut self) -> Option<()> {
-        self.line_buffer.redo()
+        if self.index_undo > 2 {
+            self.index_undo = self.index_undo.checked_sub(2)?;
+            self.undo()
+        } else {
+            None
+        }
     }
 
-    pub fn reset_olds(&mut self) {
-        self.line_buffer.reset_olds()
+    pub fn undo(&mut self) -> Option<()> {
+        // NOTE: Leaving the old logic to help with the review
+        // self.lines = self.old_lines.get(self.get_index_undo())?.clone();
+        // let insertion_point = *self.old_insertion_point.get(self.get_index_undo())?;
+        // self.set_insertion_point(insertion_point.line, insertion_point.offset);
+
+        self.line_buffer = self.edits.get(self.get_index_undo())?.clone();
```

Comment:
It does a bit of simplification IMO. (Instead or 2 arrays we only deal with one).

@sholderbach (https://discord.com/channels/601130461678272522/855886335980994600/868896677883420733) I think we can push it a bit more and have a list-zipper that way we will not even need to have the index tracking field. 
IMO ListZipper will remove the possibility of getting the index into an inconsistent state. Though I don't really know of any rust libs which have a list zipper.

(P.S. - I am a huge proponent of making impossible states unrepresentable, If I am going too far in this direction please do point it out)

Reply:
How should the API for a list zipper look like in rust? Enum with legal tuples or Struct with datastructure and index encapsulated with the legal transitions as methods?

Reply:
### Defining the List Zipper
```rust
struct ListZipper<A> {
    prev: Stack<A>,
    current: A,
    next: Stack<A>
}
```
This is the formulation that I had in mind. 
Note: This requires at-least one element to work. We can obviously use an `Option<A>` for the `current` field so that it can be empty too.

It can now support operations like 
`back` -> which `push`es `current` onto `prev` and `pop`s top from `next` to `current`
`forward` -> which `push`es `current` onto `next` and `pop`s top from `prev` to `current`

### Usage in relation to undo-redo
And `undo` basically corresponds to going `back` and redo corresponds to going `forward`.
Another API we would need is invalidating the `next` `Stack` when a user `undo`s and then starts typing new stuff.

> Enum with legal tuples or Struct with datastructure and index encapsulated with the legal transitions as methods?
Note: Given this structure, we should need an index 

### Pros
- State management is not happening in 2 places which can go out of sync
- No need for addition/subtraction hacks that we are doing currently in order to handle cases where index rules fail etc
- We don't really need the version which works with empty. So no need to handle the `Option<A>` field

### Cons
- There is no crate that has this (?), So having to implement this doesn't seem worthwhile enough to me (given it is just used in one place, if there was a lib that we could leverage then I think it would have been something worth exploring)

Reply:
This struct could also be a simple wrapper over `Vec` and `index` with accessors and the valid operations, that we implement and test separately. If we consider that performance critical, this would probably suffice. The undo stack is most likely not a hot path and wouldn't require cache locality compared to the history where similar movement is performed but we want the search to be fast.

Reply:
> This struct could also be a simple wrapper over Vec and index with accessors and the valid operations, that we implement and test separately.

Yes, that will work too. Though we do miss out on the "impossible states being unrepresentable" bit, which is sort of the main argument in favour of using a List Zipper.

I want to reiterate that I would be fine with doing it the present way too where the Editor maintains the pointer. Moreover in the particular PR, I would suggest that we don't involve the ListZipper. And if in future the Editor becomes unwieldy then we can extract this abstraction out. 
The major reason I wanted to use ListZipper was it simplifies API + it is something that I assume existed and we would only have to import the lib. But given the lib does not exist I don't really find it worthwhile to have this abstraction. Given that Editor is simple enough right now.
Thoughts?

Reply:
My two cents: zippers are a cool technique, but I think I'm hesitant to use things like that because it assumes that other contributors will be familiar with them. The only context I've seen them used is functional programming, and I think not many people come from a functional programming background. Similar to how monads would assume functional programming background.

It's okay if a technique isn't "perfect" when it's easy for most contributors to go in, understand how it works quickly, and can fix issues with it pretty quickly.

Again, just my two cents.